relentless/ensemble.py

```diff
@@ -1,176 +1,437 @@
-            if 'mu_{}'.format(t) in self.conjugates:
-                self._mu[t] = value[t]
-            else:
-                raise AttributeError('Chemical potential is not a conjugate variable.')
+    @property
+    def types(self):
+        """tuple: The types in the ensemble."""
+        return self.N.keys
 
     @property
-    def conjugates(self):
-        return self._conjugates
+    def constant(self):
+        """dict: The constant variables in the system.
+
+        No changse should be made to this dictionary, as the constant properties
+        are fixed at construction.
 
-    def reset(self):
-        if 'V' in self.conjugates:
+        """
+        return self._constant
+
+    def clear(self):
+        """Clears all conjugate variables and all rdf values in the ensemble
+        (set to `None`).
+
+        Returns
+        -------
+        :py:class:`Ensemble`
+            Returns the ensemble object with cleared parameters.
+
+        """
+        if not self.constant['V']:
             self._V = None
-        if 'P' in self.conjugates:
+        if not self.constant['P']:
             self._P = None
         for t in self.types:
-            if 'N_{}'.format(t) in self.conjugates:
+            if not self.constant['N'][t]:
                 self._N[t] = None
-            if 'mu_{}'.format(t) in self.conjugates:
+            if not self.constant['mu'][t]:
                 self._mu[t] = None
-        self.rdf = core.PairMatrix(self.types)
+        for pair in self.rdf:
+            self.rdf[pair] = None
 
         return self
 
     def copy(self):
-        ens = copy.deepcopy(self)
-        return ens.reset()
-
-    def save(self, basename='ensemble'):
-        # dump thermo data to json file
-        data = {'types': self.types,
-                'kB': self.kB,
-                'T': self.T,
-                'P': self.P,
-                'V': self.V,
-                'N': self.N,
-                'mu': self.mu,
-                'conjugates': self.conjugates}
-        with open('{}.json'.format(basename),'w') as f:
-            json.dump(data, f, sort_keys=True, indent=4)
-
-        # dump rdfs in separate files
-        for pair in self.rdf:
-            if self.rdf[pair] is not None:
-                i,j = pair
-                np.savetxt('{}.{}.{}.dat'.format(basename,i,j), self.rdf[pair].table, header='r g[{},{}](r)'.format(i,j))
+        """Makes a copy of the ensemble.
+
+        Returns
+        -------
+        :py:class:`Ensemble`
+            A copy of the ensemble object.
+
+        """
+        return copy.deepcopy(self)
+
+    def save(self, filename):
+        """Saves the ensemble parameters and the RDF values into a JSON file.
+
+        Parameters
+        ----------
+        filename : `str`
+            The name of the file to save data in.
+
+        """
+        data = []
+
+        # append thermo data
+        thermo = {'T': self.T,
+                  'P': self.P,
+                  'V': self.V,
```

Comment:
Ah, now we do have a problem (like we had with the Variables) because Volume is a class so we will need to add a method to serialize / deserialize it. Let's add abstract `to_json` instance method and `from_json` class method to Volume. The `to_json` method should return a dict containing whatever arguments are required to reconstruct it. The `from_json` method should take a dict, parse it and construct a new instance of the object, and return the object.

Then, when you are dumping the volume, I would save it as a two part dictionary:
```py
data['V'] = {'__name__' : self.V.__name__,
             'data': self.V.to_json()}
```
so you can use the name in the load. I think something like this would work for that:
```py
self.V = globals()[data['V']['__name__']].from_json(data['V']['data'])
```
You will need to implement these methods for each of the volume classes.